File: sdk/python/easynet_sdk/_session_authority_subjects.py
"""Session-authority subject admission helpers for canonical runtime calls."""

from __future__ import annotations

from .authority import SessionAuthority
from .axon_addressing import AddressingProjection, parse_ura, resource_ura, user_ura
from .errors import SDKError
from ._identity_guards import contains_all_zero_principal

_RUNTIME_STATE_READ_SUBJECT_PATH = "runtime-state/read"


def runtime_state_read_subject_ura(realm: str, user_id: str) -> str:
    """Build the canonical user-owned subject for runtime-state reads."""

    clean_realm = str(realm).strip()
    clean_user_id = str(user_id).strip()
    if not clean_realm:
        _invalid_runtime_state_subject("runtime-state read subject realm is required")
    if not clean_user_id:
        _invalid_runtime_state_subject("runtime-state read subject user_id is required")
    if contains_all_zero_principal(clean_user_id):
        _invalid_runtime_state_subject(
            "runtime-state read subject user_id must not be all-zero"
        )
    try:
        owner_ura = user_ura(clean_realm, clean_user_id)
        subject = resource_ura(owner_ura, _RUNTIME_STATE_READ_SUBJECT_PATH)
    except SDKError as error:
        _invalid_runtime_state_subject(
            "runtime-state read subject_ura must be canonical", error
        )
    return subject


def is_runtime_state_read_subject_ura(subject_ura: str) -> bool:
    try:
        subject = parse_ura(str(subject_ura).strip())
    except SDKError:
        return False
    owner_id = subject.components.get("owner_id")
    path = subject.components.get("path")
    return (
        subject.kind == "resource"
        and isinstance(owner_id, str)
        and owner_id.startswith("user.")
        and owner_id.removeprefix("user.").strip() != ""
        and not contains_all_zero_principal(owner_id.removeprefix("user."))
        and path == _RUNTIME_STATE_READ_SUBJECT_PATH
    )


def _invalid_runtime_state_subject(message: str, cause: Exception | None = None) -> None:
    raise SDKError(
        code="INVALID_INVOCATION",
        stage="authority",
        retry="never",
        retryable=False,
        message=message,
        cause=cause,
    )


def session_authority_admits_subject(
    authority: SessionAuthority,
    subject_ura: str,
    subject: AddressingProjection | None = None,
) -> bool:
    """Return whether a session authority admits one descriptor-bound subject.

    The rule mirrors the Go SDK helper:
    - exact `subject_ura` equality is always accepted;
    - a user session authority may admit resources owned by that user;
    - an agent-owned resource is admitted only when the agent owner is the same
      user.

    Path substring matches are intentionally not accepted. Ownership must come
    from the canonical URA projection's `owner_id` component.
    """

    if authority.subject_ura.strip() == subject_ura.strip():
        return True
    if subject is None:
        try:
            subject = parse_ura(subject_ura.strip())
        except SDKError:
            return False
    if subject.kind != "resource":
        return False
    owner_id = subject.components.get("owner_id")
    if not isinstance(owner_id, str):
        return False
    owner_user_id = authority.session_owner_user_id.strip()
    if not owner_user_id:
        return False
    if owner_id == f"user.{owner_user_id}":
        return True
    if not owner_id.startswith("agent."):
        return False
    agent_owner = owner_id.removeprefix("agent.").split(".", 1)
    return len(agent_owner) == 2 and agent_owner[0] == owner_user_id

File: sdk/go/session_authority_subjects.go
package easynet

import "strings"

const runtimeStateReadSubjectPath = "runtime-state/read"

// RuntimeStateReadSubjectURA builds the canonical subject for runtime-state
// read projections owned by one authenticated user.
//
// Products pass realm/user identity into this helper instead of defaulting
// read-only invocations to the target device. Admission still validates the
// resulting tuple; this helper only centralizes the product-neutral subject
// projection used by history, catalogue, and status reads.
func RuntimeStateReadSubjectURA(realm string, userID string) (string, error) {
	realm = strings.TrimSpace(realm)
	userID = strings.TrimSpace(userID)
	if realm == "" {
		return "", invalidInvocation("runtime-state read subject realm is required", nil)
	}
	if userID == "" {
		return "", invalidInvocation("runtime-state read subject user_id is required", nil)
	}
	if containsAllZeroPrincipal(userID) {
		return "", invalidInvocation("runtime-state read subject user_id must not be all-zero", nil)
	}
	subject := ResourceDotURA(realm, "user."+userID, runtimeStateReadSubjectPath)
	if _, err := ParseURA(subject); err != nil {
		return "", invalidInvocation("runtime-state read subject_ura must be canonical", err)
	}
	return subject, nil
}

func isRuntimeStateReadSubjectURA(subjectURA string) bool {
	parts, err := ParseURAParts(strings.TrimSpace(subjectURA))
	if err != nil || parts.Kind != URAKindResource {
		return false
	}
	ownerID := strings.TrimSpace(parts.OwnerID)
	userID := strings.TrimPrefix(ownerID, "user.")
	return strings.HasPrefix(ownerID, "user.") &&
		strings.TrimSpace(userID) != "" &&
		!containsAllZeroPrincipal(userID) &&
		strings.TrimSpace(parts.Path) == runtimeStateReadSubjectPath
}

// runtimeSessionAuthorityAdmitsSubject is the Go SDK's canonical
// session-authority subject admission predicate. Runtime ability calls and
// invocation-history queries both consume this helper; neither path owns a
// private subject expansion rule.
func runtimeSessionAuthorityAdmitsSubject(
	authority *SessionAuthority,
	subjectURA string,
) bool {
	if authority == nil {
		return false
	}
	if strings.TrimSpace(authority.SubjectURA) == strings.TrimSpace(subjectURA) {
		return true
	}
	parts, err := ParseURAParts(strings.TrimSpace(subjectURA))
	if err != nil || parts.Kind != URAKindResource {
		return false
	}
	ownerID := strings.TrimSpace(parts.OwnerID)
	ownerUserID := strings.TrimSpace(authority.SessionOwnerUserID)
	if ownerUserID == "" {
		return false
	}
	if strings.HasPrefix(ownerID, "user.") && strings.TrimPrefix(ownerID, "user.") == ownerUserID {
		return true
	}
	if !strings.HasPrefix(ownerID, "agent.") {
		return false
	}
	agentOwner := strings.TrimPrefix(ownerID, "agent.")
	userID, _, found := strings.Cut(agentOwner, ".")
	return found && userID == ownerUserID
}
